Document update/change/delete in block format incremental import (#2177)
neo4j/docs-operations: modules/ROOT/pages/tools/neo4j-admin/neo4j-admin-import.adoc

----
:ACTION,:START_ID,:TYPE,:END_ID,p1{identifier:true},name,p4
U,1,KNOWS,2,abc,"Keanu Reeves","Hello Morpheus"
U,1,KNOWS,2,def,"Laurence Fishburne","Hello Neo"
----

The data in the `p1` column for these relationships acts "more uniquely" selecting relationships, if multiple of `1,KNOWS,2` exist.
Identifier properties will match the selected relationships and as such not be set on the relationships (they already have it).

=== Update existing labels

You can add or remove one or more labels from an existing node by prepending the clause `LABEL` in the header with a `+` (default) or `-`:

* `:+LABEL` - Add one or more labels to an existing node.
* `:-LABEL` - Remove one or more labels (if they exist) from an existing node.

Multiple labels are separated by a semicolon (default).

For example, a file could have the following format:

[source, csv]
----
:ID,:+LABEL,:-LABEL,name,age
person1,Actor,Person,"Keanu Reeves",55
person2,Actor;Director,Person,"Laurence Fishburne",60
person3,Actor,Person,"Carrie-Anne Moss",53
----

In this case, all labels in the second column are added and all the labels in the third column are removed (if they exist).

=== Remove existing properties

> You can remove properties from existing nodes or relationships by specifying them after the keyword `:-PROPERTY` in the header file.

Reviewer: Suggestion: "You can remove properties from existing nodes or relationships by adding a `:-PROPERTY` column in the header. In the contents of this field you can add zero or more property names to remove from the entity".

> You can remove properties from existing nodes or relationships by specifying them after the keyword `:-PROPERTY` in the header file.

Author: ```suggestion
You can remove properties from existing nodes or relationships by a `:-PROPERTY` column in the header.
In the contents of this field you can add zero or more property names to remove from the entity.
```

All nodes are identified by their `:ID` and all relationships by their `:START_ID`, `:END_ID`, and `:TYPE`.
For example:

.Remove nodes' properties
[source, cypher, role="nocopy"]
----

> :ID,:LABEL,:-PROPERTY,age,hometown

Author: Ok, so in this case:
```
:ID,:LABEL,:-PROPERTY,age,hometown
person1,Actor,55,New York
```
it will remove some property (which I don't know what it is), and we leave the rest.
Is that the case?
Also in your example, there is no label but further up, you said that nodes are identified by their unique property value for the key/label combination that the header specifies. Doesn't this mean that the header should contain a label as well?

> :ID,:LABEL,:-PROPERTY,age,hometown

Reviewer: Your example should be:

```
:ACTION,:ID,:-PROPERTY
U,person1,age;hometown
```

and then a more meaningful example would probably state the full capabilities of update, i.e. to add/remove labels and set/remove properties all at the same time (I'm not good at examples):

```
:ACTION,uid:ID(label:Person),:LABEL,:-LABEL,:-PROPERTY,name,height:int
U,person1,Actor,Producer,age;hometown,"Henry",185
```

(And also the usual way of identifying nodes in incremental still apply, i.e. you need to specify e.g. `uid` and `Person` to let it know which uniqueness constraint you're looking up the nodes in.

See, one CSV entry can specify all types of updates to one entity at the same time. Here the node `person1` is updated with:
- added `Actor` label
- removed `Producer` label
- removed `age` and `hometown` properties
- set `name="Henry"` property
- set `height=185` property

> :ID,:LABEL,:-PROPERTY,age,hometown

Author: ```suggestion
:ACTION,:ID,:LABEL,:-PROPERTY,age,hometown
```

> person1,Actor,55,New York

Author: ```suggestion
U,person1,Actor,55,New York
```

----

> Properties `age` and `hometown` are removed from the node with the `:ID` `person1` and label `Actor`.

Author: ```suggestion
Properties `55` and `New York` are removed from the node with the `:ID` `person1` and label `Actor`.
```

.Remove relationships' properties
[source, cypher, role="nocopy"]
----

> :START_ID,:END_ID,:TYPE,:-PROPERTY, role, description

Reviewer: This example has the same problem as above

> :START_ID,:END_ID,:TYPE,:-PROPERTY, role, description

Author: ```suggestion
:ACTION, :START_ID,:END_ID,:TYPE,:-PROPERTY, role, description
```

> 1,2,ACTED_IN,role,description

Author: ```suggestion
1,2,ACTED_IN,Neo,"The chosen one"
```

----

> Properties `role` and `description` are removed from the relationship with the `:START_ID` `1`, `:END_ID` `2`, and `:TYPE` `ACTED_IN`.

Author: ```suggestion
What is remove here if not `role`and `description`
```
```suggestion
Properties `Neo`, `"The chosen one"` are removed from the relationship with the `:START_ID` `1`, `:END_ID` `2`, and `:TYPE` `ACTED_IN`.
```

CREATE,1,2,ACTED_IN,"Neo"
UPDATE,1,2,ACTED_IN,"Morpheus"
DELETE,1,3,ACTED_IN
----

Relationships are identified non-uniquely by their start and end node IDs, and their type.

To further narrow down selection you can tag a property column as an identifier to help out in selecting relationships uniquely (or at least more uniquely).

.Using actions in CSV files to update relationships with identifier properties
[source, cypher, role="nocopy"]
----
:ACTION,:START_ID,:TYPE,:END_ID,p1{identifier:true},name,p4
U,1,KNOWS,2,abc,"Keanu Reeves","Hello Morpheus"
U,1,KNOWS,2,def,"Laurence Fishburne","Hello Neo"
----

The data in the `p1` column for these relationships acts "more uniquely" selecting relationships, if multiple of `1,KNOWS,2` exist.
Identifier properties will match the selected relationships and as such not be set on the relationships (they already have it).

=== Update existing labels

You can add or remove one or more labels from an existing node by prepending the clause `LABEL` in the header with a `+` (default) or `-`:

* `:+LABEL` - Add one or more labels to an existing node.
* `:-LABEL` - Remove one or more labels (if they exist) from an existing node.

> Multiple labels are separated by a semicolon (default).

Author: ```suggestion
```

For example, a file could have the following format:

[source, csv]
----
:ID,:+LABEL,:-LABEL,name,age
person1,Actor,Person,"Keanu Reeves",55
person2,Actor;Director,Person,"Laurence Fishburne",60
person3,Actor,Person,"Carrie-Anne Moss",53
----

In this case, all labels in the second column are added and all the labels in the third column are removed (if they exist).

=== Remove existing properties

You can remove properties from existing nodes or relationships by a `:-PROPERTY` column in the header.
In the contents of this field you can add zero or more property names to remove from the entity.
All nodes are identified by their `:ID` and all relationships by their `:START_ID`, `:END_ID`, and `:TYPE`.
For example:

.Remove nodes' properties
[source, cypher, role="nocopy"]
----
:ACTION,:ID,:-PROPERTY
U,person1,Actor,55,New York
----

Properties `55` and `New York` are removed from the node with the `:ID` `person1`.

.Remove relationships' properties
[source, cypher, role="nocopy"]
----
:ACTION, :START_ID,:END_ID,:TYPE,:-PROPERTY

> U, 1,2,ACTED_IN,Neo,"The chosen one"

Author: ```suggestion
U,1,2,ACTED_IN,Neo,"The chosen one"
```

----

> Properties `role` and `description` are removed from the relationship with the `:START_ID` `1`, `:END_ID` `2`, and `:TYPE` `ACTED_IN`.

Author: ```suggestion
Properties Neo` and `"The chosen one"` are removed from the relationship with the `:START_ID` `1`, `:END_ID` `2`, and `:TYPE` `ACTED_IN`.
```

Deleting a node also deletes its relationships (`DETACH DELETE`).

.Using actions in CSV files to update nodes
[source, cypher, role="nocopy"]
----
:ACTION,uid:ID(label:Person),name,:LABEL
CREATE,person1,"Keanu Reeves",Actor
UPDATE,person2,"Laurence Fishburne",Actor
DELETE,person4,, 
----

Nodes are identified by their unique property value for the key/label combination that the header specifies.

.Using actions in CSV files to update relationships
[source, cypher, role="nocopy"]
----
:ACTION,:START_ID,:END_ID,:TYPE,role
CREATE,person1,movie1,ACTED_IN,"Neo"
UPDATE,person2,movie1,ACTED_IN,"Morpheus"
DELETE,person3,movie1,ACTED_IN
----

Relationships are identified non-uniquely by their start and end node IDs, and their type.

To further narrow down selection you can tag a property column as an identifier to help out in selecting relationships uniquely (or at least more uniquely).

.Using actions in CSV files to update relationships with identifier properties
[source, cypher, role="nocopy"]
----
:ACTION,:START_ID,:TYPE,:END_ID,p1{identifier:true},name,p4
U,person1,KNOWS,person2,abc,"Keanu Reeves","Hello Morpheus"
U,person2,KNOWS,person1,def,"Laurence Fishburne","Hello Neo"
----

> The data in the `p1` column for these relationships helps "more uniquely" selecting relationships, if multiple of `1,KNOWS,2` exist.

Author: ```suggestion
The data in the `p1` column for these relationships helps select relationships "more uniquely" if a multiple of `1,KNOWS,2` exists.
```

> Identifier properties will match the selected relationships and as such not be set on the relationships (they already have it).

Author: ```suggestion
Identifier properties match the selected relationships and will not be set on the relationships that already have them.
```

====

[role=label--new-2025.01.0]
== Applying changes to data via CSV files

You can use CSV files to update existing nodes, relationships, labels, or properties during incremental import.

[NOTE]
====
This feature is supported only by `block` format.
====

=== Set an explicit action for each row

You can set an explicit action for each row in the CSV file by using the `:ACTION` keyword in the header file.
If no action is specified, the import tool works as in full import mode, creating new data.

The following actions are supported:

* `empty` = `CREATE` (default)
* `C`, `CREATE` - Creates new nodes and relationships, with or without properties, as well as labels.
* `U`, `UPDATE` - Updates existing nodes, relationships, labels, and properties.
* `D`, `DELETE` - Deletes existing nodes or relationships.
Deleting a node also deletes its relationships (`DETACH DELETE`).

.Using actions in CSV files to update nodes
[source, cypher, role="nocopy"]
----
:ACTION,uid:ID(label:Person),name,:LABEL
CREATE,person1,"Keanu Reeves",Actor
UPDATE,person2,"Laurence Fishburne",Actor

> DELETE,person4

Author: Interesting, this is from the example you gave me, but I can't find it anymore.

> DELETE,person4

Author: ```suggestion
DELETE,person4, , 
```

> DELETE,person4

Reviewer: Yeah, I mean it's fine as it is. It's just not 100% correct CSV structure :) probably my fault

----

Nodes are identified by their unique property value for the key/label combination that the header specifies.

.Using actions in CSV files to update relationships
[source, cypher, role="nocopy"]
----
:ACTION,:START_ID,:END_ID,:TYPE,role
CREATE,person1,movie1,ACTED_IN,"Neo"
UPDATE,person2,movie1,ACTED_IN,"Morpheus"
DELETE,person3,movie1,ACTED_IN
----

Relationships are identified non-uniquely by their start and end node IDs, and their type.

To further narrow down selection you can tag a property column as an identifier to help out in selecting relationships uniquely (or at least more uniquely).

.Using actions in CSV files to update relationships with identifier properties
[source, cypher, role="nocopy"]
----
:ACTION,:START_ID,:TYPE,:END_ID,p1{identifier:true},name,p4
U,person1,KNOWS,person2,abc,"Keanu Reeves","Hello Morpheus"
U,person2,KNOWS,person1,def,"Laurence Fishburne","Hello Neo"
----

> The data in the `p1` column for these relationships acts "more uniquely" selecting relationships, if multiple of `1,KNOWS,2` exist.

Author: ```suggestion
The data in the `p1` column for these relationships helps "more uniquely" selecting relationships, if multiple of `1,KNOWS,2` exist.
```

Identifier properties will match the selected relationships and as such not be set on the relationships (they already have it).

=== Update existing labels

You can add or remove one or more labels from an existing node by prepending the clause `LABEL` in the header with a `+` (default) or `-`:

* `:+LABEL` - Add one or more labels to an existing node.
* `:-LABEL` - Remove one or more labels (if they exist) from an existing node.

For example, a file could have the following format:

[source, csv]
----
uid:ID(label:Person),:+LABEL,:-LABEL,name,age
person1,Actor,Producer,"Keanu Reeves",55

> person2,Actor;Director,"Laurence Fishburne",60

Author: Ah, yes.

> person2,Actor;Director,"Laurence Fishburne",60

Author: ```suggestion
person2,Actor;Director,,"Laurence Fishburne",60
```

----

In this case, all labels in the second column are added and all the labels in the third column are removed (if they exist).

=== Remove existing properties

You can remove properties from existing nodes or relationships by a `:-PROPERTY` column in the header.
In the contents of this field you can add zero or more property names to remove from the entity.

> All nodes are identified by their `:ID` and all relationships by their `:START_ID`, `:END_ID`, and `:TYPE`.

Author: yes, it was, but I'll remove it.

> All nodes are identified by their `:ID` and all relationships by their `:START_ID`, `:END_ID`, and `:TYPE`.

Author: ```suggestion
```

For example:

.Remove nodes' properties
[source, cypher, role="nocopy"]
----
:ACTION,uid:ID(label:Person),:-PROPERTY

> U,person1,55;New York

Author: ```suggestion
U,person1,age;hometown
```

----

> Properties `55` and `New York` are removed from the node with the `uid:ID` `person1`.

Reviewer: Shouldn't it say that it's properties `age` and `hometown` are removed?

> Properties `55` and `New York` are removed from the node with the `uid:ID` `person1`.

Author: ```suggestion
Properties `age` and `hometown` are removed from the node with the `uid:ID` `person1`.
```

.Remove relationships' properties
[source, cypher, role="nocopy"]
----
:ACTION,:START_ID,:END_ID,:TYPE,:-PROPERTY

> U,person1,movie1,ACTED_IN,Neo;"The chosen one"

Reviewer: Likewise here: `Neo;"The chosen one"` are property values. The contents of this field should be property names, e.g. `name;role` or something like that.

----

> Properties `Neo` and `"The chosen one"` are removed from the relationship with the `:START_ID` `person1`, `:END_ID` `movie1`, and `:TYPE` `ACTED_IN`.

Reviewer: Shouldn't it say that properties `role` and `description` are removed?

Author: Ah, yes, thanks.

> Properties `Neo` and `"The chosen one"` are removed from the relationship with the `:START_ID` `person1`, `:END_ID` `movie1`, and `:TYPE` `ACTED_IN`.

Author: ```suggestion
Properties `role` and `description` are removed from the relationship with the `:START_ID` `person1`, `:END_ID` `movie1`, and `:TYPE` `ACTED_IN`.
```

.Using actions in CSV files to update labels and properties
[source, cypher, role="nocopy"]
----
:ACTION,uid:ID(label:Person),:LABEL,:-LABEL,:-PROPERTY,name,height:int

> U,person1,Actor,Producer,46;New York,"Henry",185

Reviewer: Same problem with `:-PROPERTY` here, i.e. `45;New York` should instead be `age;hometown`

> U,person1,Actor,Producer,46;New York,"Henry",185

Author: ```suggestion
U,person1,Actor,Producer,age;hometown,Henry",185
```

The `id` property of the nodes in the `persons` group will be stored as `long` type, while the `id` property of the nodes in the `games` group will be stored as `string` type, as the global `id-type` is a string.
====

[role=label--new-2025.01.0 label--block-only]
== Applying changes to data via CSV files

You can use CSV files to update your data during incremental import.

> This feature is especially useful when you want to update existing nodes, relationships, labels, or properties in large datasets.

Author: What do you suggest?

To be able to use this feature, each entity in the dataset must be associated with a unique ID.

[NOTE]
====
This feature is supported only by `block` format.
====

=== Set an explicit action for each row

You can set an explicit action for each row in the CSV file by using the `:ACTION` keyword in the header file.
If no action is specified, the import tool works as in full import mode, creating new data.

The following actions are supported:

* `empty` = `CREATE` (default)
* `C`, `CREATE` - Creates new nodes and relationships with or without properties.
* `U`, `UPDATE` - Updates existing nodes, relationships, or properties.

> Leaving the property identifier empty is like having a “*” wild card.

Author: This I saw it in the PRD where I was looking for information. This was the example:

```
:START_ID,:END_ID,:TYPE,prop1:int{identifier=true},prop2:string{identifier=true},prop3,:ACTION
0        ,1      ,R    ,0                         ,Hi                           ,Bob  ,UPDATE
0        ,1      ,R    ,1                         ,                             ,Joe  ,UPDATE
``` 
Leaving the identifier property empty is like having a “*” wild card.

* `D`, `DELETE` - Deletes an existing node or relationship.
Deleting a node also deletes its relationships (`DETACH DELETE`).

.Using actions in CSV files to update data
[source, cypher, role="nocopy"]
----
:ACTION,personId:ID,name,:LABEL
CREATE,1,"Keanu Reeves",Actor
UPDATE,2,"Laurence Fishburne",Actor
DELETE,4
----

=== Update existing labels

You can add or remove one or more labels from an existing node by prepending the clause `LABEL` in the header with a a `+` (default) or `-`:

* `:+LABEL` - Add one or more labels to an existing node.
* `:-LABEL` - Remove one or more labels (if they exist) from an existing node.

> Multiple labels are separated by a semicolon.

Author: I couldn't find it anywhere on this page.

> Multiple labels are separated by a semicolon.

Author: And I also didn't know that the semicolon for array delimiter is the default. Why do we need to mention that it can be configured using a command option given that it's already configured to be the default?

> Multiple labels are separated by a semicolon.

Reviewer: Because if the manual mentions that it's a semicolon that is the delimiter then it looks like that's that, right. Better to perhaps talk about "array delimiter character (which defaults to semicolon)" to be clear on that!

For example, a file could have the following format:

[source, csv]
----
p0:ID,:+LABEL,:-LABEL,name,age
person1,Actor,Person,"Keanu Reeves",55
person2,Actor;Director,Person,"Laurence Fishburne",60
person3,Actor,Person,"Carrie-Anne Moss",53
----

In this case, all labels in the second column are added and all the labels in the third column are removed (if they exist).
The nodes are identified by their `p0` key.

=== Update existing properties

You can add or remove properties from an existing entity by prepending the clause `PROPERTY` in the header with a `+` or `-`:

> * `:+PROPERTY` - Add one or more properties (if they exist) to an existing entity.

Reviewer: This isn't how it works. `:+PROPERTY` doesn't exist. You specify properties using their name in the column header.

* `:-PROPERTY` - Remove one or more properties (if they exist) from an existing entity.

> Multiple properties are separated by a semicolon.

Reviewer: Also, same here about semicolon simply being the default.
